javascript/extractor/lib/typescript/BUILD.bazel

```diff
@@ -0,0 +1,55 @@
+load("@//:common.bzl", "on_windows")
+
+# Builds a zip file of the compiled typscript-parser-wrapper and its dependencies.
+genrule(
+    name = "typescript",
+    srcs = [
+        "tsconfig.json",
+        "package.json",
+        "package-lock.json",
+    ] + glob(["src/**"]),
+    outs = ["javascript.zip"],
+    cmd = " && ".join([
+        # the original working directory is not preserved anywhere, but needs to be accessible, as
+        # all paths are relative to this
+        # unfortunately, we need to change the working directory to run npm.
+        "export BAZEL_ROOT=$$(pwd)",
+        # we need a temp dir, and unfortunately, $TMPDIR is not set on Windows
+        "export TEMP=$$(mktemp -d)",
+        # Add node to the path so that npm run can find it - it's calling env node
+        "export PATH=$$PATH:$$BAZEL_ROOT/$$(dirname $(execpath @nodejs//:node_bin))",
+        "export NPM=$$BAZEL_ROOT/$(execpath @nodejs//:npm_bin)",
+        # npm has a global cache which doesn't work on macos, where absolute paths aren't filtered out by the sandbox.
+        # Therefore, set a temporary cache directory.
+        "export NPM_CONFIG_USERCONFIG=$$TEMP/npmrc",
+        "$$NPM config set cache $$TEMP/npm",
+        "$$NPM config set update-notifier false",
+        "rm -rf $(RULEDIR)/inputs",
+        "cp -L -R $$(dirname $(execpath package.json)) $(RULEDIR)/inputs",
+        "cd $(RULEDIR)/inputs",
+        "$$NPM install",
+        "$$NPM run build",
+        "rm -rf node_modules",
+        # Install again with only runtime deps
+        "$$NPM install --prod",
+        "mv node_modules build/",
+        "mkdir -p javascript/tools/typescript-parser-wrapper",
+        "mv build/* javascript/tools/typescript-parser-wrapper",
+        "",
+    ]) + on_windows(
+        " && ".join([
+            "$$BAZEL_ROOT/$(execpath @bazel_tools//tools/zip:zipper) cC $$(cygpath -w $$BAZEL_ROOT/$@) $$(find javascript -name '*' -print)",
+            "rm -rf $$TEMP",
+        ]),
+        " && ".join([
+            "$$BAZEL_ROOT/$(execpath @bazel_tools//tools/zip:zipper) cC $$BAZEL_ROOT/$@ $$(find javascript -name '*' -print)",
+            "rm -rf $$TEMP",
+        ]),
```

Comment:
when genrules become _this_ complex, it might make sense to move everything to a helper `sh_binary` or `py_binary` doing all this heavy lifting with just some parameters or runfiles provided to it. But I understand if you want to leave it for later if we ever need to touch this again

Reply:
Yeah, this makes sense.

I tried converting the test runner to python due to precisely this. It would be nice, as then we could sidestep all the gnarly msys-path issues on Windows.
However, the python environment under bazel is slightly off:
* All dependencies get zipped up (including that nodejs/npm we're depending on), taking some extra time
* At runtime, the paths of the external repositories don't match either `$(rootpath )` or `$(execpath )` - instead of being available at `<cwd>/external/<dep>`, external dependencies are at `<cwd>/../<dep>`.

Thus, if you take in paths to your dependencies as arguments to your script, you need to hardcode that path transformation - pretty ugly. I guess the alternative would be pulling in a runfiles library to resolve these paths, instead of resolving them in the arguments to the script? That seems just above the level of complication needed for a 20line python script.
I also tried using the bash runfiles library for the test runner, and that seems to be unable to resolve anything that's not sh_librarys - external tool dependencies didn't work. Ergo, back to genrule/sh_test it is.

Why the path setup isn't the same as that of a genrule honestly escapes me, but I assume it's some legacy reason.

Reply:
this picked my interest, so I tried out the `sh_binary` route. I agree that it is far from a nice experience. I think the biggest problem is that the runfiles libraries deal with actual file paths instead of bazel labels. So to get the location of `zipper` or `npm` you need to know the exact file path, that

a. is not easily known when one writes this
b. it may even change with respect to the platform (`npm` for example is within `nodejs_linux_amd64` on my platform)

It'd be much easier if the runfile manifest file also contained the mapping from bazel targets, so that one could use with the runfiles library something like `$(rlocation @bazel_tools//tools/zip:zipper)` instead of having to spell out the full path.

That said, here's [a shot at it](https://github.com/github/codeql/pull/14730) bypassing that issue with a `find` and an `$(rlocation .)` to give the runfiles root directory. You still need to know the filenames of the tools, and I don't find it necessarily better than what you did with `genrule`. So take it as an experiment, and I'll leave it to your judgement whether it's worth going that way.

Reply:
<img width="122" alt="Screenshot 2023-11-09 at 10 16 09" src="https://github.com/github/codeql/assets/70492/f2285d7a-297f-476c-820a-c99b9cb8e12f">

says it all, imo.
I agree that runfiles with the labels ought to work, but it doesn't, and that makes it a not-so-great solution.

Reply:
yeah, I see we arrived at the same conclusions (I did not see your further comment while I was writing mine).

> Besides, the bash runfiles lib accepts paths in its own syntax, not bazel targets 🤯

why, other runfiles libraries do? I don't think so, I don't think the runfiles manifest contains that information in general. But I've started [this discussion](https://github.com/bazelbuild/bazel/discussions/20117) for that.

Reply:
I'm not sure, I've never used a runfiles lib before. I guess I had just assumed that it would be able to take bazel targets

Reply:
me too :shrug:

Reply:
> <img alt="Screenshot 2023-11-09 at 10 16 09" width="122" src="https://user-images.githubusercontent.com/70492/281679072-f2285d7a-297f-476c-820a-c99b9cb8e12f.png">
> says it all, imo. I agree that runfiles with the labels ought to work, but it doesn't, and that makes it a not-so-great solution.

Well, there is the advantage of having more IDE support from the bash code in a bash script than in a `BUILD.bazel` file, but then again, I do agree it's probably not worth it.